### ISSUE TITLE
[feature] Enable Custom Component Name for Container Components

### ISSUE THREAD

**Issue by ricardo-rolo** [kubeflow/pipelines#9998]
### Feature Area

/area sdk

### What feature would you like to see?

For creating a Custom Container Component Factory, it would be nice to be able to customize the component name, without having to change the python function name.

### What is the use case or pain point?

In a Custom Container Component Factory, where one single Container Component function with the decorator defines the general way to create a component, calls different modules that compose the Components itself. That way it will be possible to have just one Container Component definition creating Components with a given name, and not with the name of the Factory Function. Ex:

```python
@container_component(name="Training")
def generic_component(
    component_class_path: str,
    image: str,
    outputs: Output[Artifact],
):
    return ContainerSpec(
        image=image,
        command=[
            "python",
            "-m",
            "components.launcher",
        ],
        args=[
            "--component_class_path",
            component_class_path,
            "--outputs",
            outputs.path,
        ],
    )

```

### Is there a workaround currently?

Basically creating one Container Component function using the decorator for each desired component.

### Suggestion: override func.\_\_name__ with a given name argument, if present, just like the standard component decorator.

[create_container_component_from_func](https://github.com/kubeflow/pipelines/blob/4bee3d8dc2ee9c33d87e1058bac2a94d899dd4a5/sdk/python/kfp/dsl/component_factory.py#L591)
[extract_component_interface](https://github.com/kubeflow/pipelines/blob/4bee3d8dc2ee9c33d87e1058bac2a94d899dd4a5/sdk/python/kfp/dsl/component_factory.py#L186)
---

Love this idea? Give it a 👍.
